src/main/flight/imu.c

```diff
@@ -201,86 +201,101 @@ static float imuUseFastGains(void) {
         }
     }
 }
+FAST_CODE void applyVectorError(float ez_ef, quaternion *vError){
+    // Rotate mag error vector back to BF and accumulate
+    vError->x += (2.0f * (qpAttitude.xz + -qpAttitude.wy)) * ez_ef;
+    vError->y += (2.0f * (qpAttitude.yz - -qpAttitude.wx)) * ez_ef;
+    vError->z += (1.0f - 2.0f * qpAttitude.xx - 2.0f * qpAttitude.yy) * ez_ef;
+}
 
-static void imuMahonyAHRSupdate(float dt, quaternion *vGyro, bool useAcc, quaternion *vAcc, bool useMag, quaternion *vMag, bool useCOG, float courseOverGround) {
-    quaternion vKpKi = VECTOR_INITIALIZE;
-    quaternion vError = VECTOR_INITIALIZE;
-    static quaternion vIntegralFB = VECTOR_INITIALIZE;
-    quaternion qBuff, qDiff;
-
+FAST_CODE void applyAccError(quaternion *vAcc, quaternion *vError) {
+    quaternionNormalize(vAcc);
+    // Error is sum of cross product between estimated direction and measured direction of gravity
+    vError->x += (vAcc->y * (1.0f - 2.0f * qpAttitude.xx - 2.0f * qpAttitude.yy) - vAcc->z * (2.0f * (qpAttitude.yz - -qpAttitude.wx)));
+    vError->y += (vAcc->z * (2.0f * (qpAttitude.xz + -qpAttitude.wy)) - vAcc->x * (1.0f - 2.0f * qpAttitude.xx - 2.0f * qpAttitude.yy));
+    vError->z += (vAcc->x * (2.0f * (qpAttitude.yz - -qpAttitude.wx)) - vAcc->y * (2.0f * (qpAttitude.xz + -qpAttitude.wy)));
+}
 
-    if (useCOG) {
-        // Use raw heading error (from GPS or whatever else)
-        while (courseOverGround >  M_PIf) courseOverGround -= (2.0f * M_PIf);
-        while (courseOverGround < -M_PIf) courseOverGround += (2.0f * M_PIf);
+FAST_CODE float calcMagError(quaternion *vMag){
+    quaternionNormalize(vMag);
+    
+    // (hx; hy; 0) - measured mag field vector in EF (assuming Z-component is zero)
```

Comment:
If you are taking my code from INAV giving some credit would be nice, don't you think?

Reply:
@digitalentity 
Karma is a bitch  (GPL)